x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/AutoscalingDecision.java

```diff
@@ -0,0 +1,220 @@
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+public abstract class AutoscalingDecision implements ToXContent, Writeable {
+
+    public abstract String name();
+
+    public abstract Type type();
+
+    public abstract String reason();
+
+    public abstract Collection<? extends AutoscalingDecision> decisions();
+
+    public static class SingleAutoscalingDecision extends AutoscalingDecision {
+
+        private final String name;
+
+        @Override
+        public String name() {
+            return name;
+        }
+
+        private final Type type;
+
+        @Override
+        public Type type() {
+            return type;
+        }
+
+        private final String reason;
+
+        @Override
+        public String reason() {
+            return reason;
+        }
+
+        @Override
+        public Collection<? extends AutoscalingDecision> decisions() {
+            return List.of(this);
+        }
+
+        public SingleAutoscalingDecision(final String name, final Type type, final String reason) {
+            this.name = Objects.requireNonNull(name);
+            this.type = Objects.requireNonNull(type);
+            this.reason = Objects.requireNonNull(reason);
+        }
+
+        public SingleAutoscalingDecision(final StreamInput in) throws IOException {
+            this.name = in.readString();
+            this.type = Type.readFrom(in);
+            this.reason = in.readString();
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeString(name);
+            type.writeTo(out);
+            out.writeString(reason);
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            builder.startObject();
+            {
+                builder.field("name", name);
+                builder.field("type", type);
+                builder.field("reason", reason);
+            }
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            final SingleAutoscalingDecision that = (SingleAutoscalingDecision) o;
+            return name.equals(that.name) && type == that.type && reason.equals(that.reason);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(name, type, reason);
+        }
+
+    }
+
+    public static class MultipleAutoscalingDecision extends AutoscalingDecision {
```

Comment:
I would prefer to not extend `AutoscalingDecision`. This could let `AutoscalingDecision` be a final class and we avoid throwing `UnsupportedOperationException` below (and remove `decisions()` from single decision).

Autoscaling deciders should not really have to choose between using a single or multi decision. The multi-decision is only for the "framework". `MultipleAutoscalingDecision` could be an inner class to `AutoscalingService` or co-located with the service.

I know this is contrary to lots of other code, so please consider this optional.

Reply:
Thanks for the input. I have made a refactoring here. Would you let me know your thoughts?

Reply:
Thanks for doing the refactor, this looks great.